### PR TITLE
net: lwm2m: Documentation of the lwm2m engine

### PR DESCRIPTION
**Pull request by otkulseng**
I've provided documentation for the API's
- lwm2m_registry.h
- lwm2m_engine.h

Signed-off-by: Ola Tangen Kulseng <ola.kulseng@nordicsemi.no>

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_engine.h [hunk 1]
```diff
@@ -30,63 +30,286 @@
 /* length of time in milliseconds to wait for buffer allocations */
 #define BUF_ALLOC_TIMEOUT K_SECONDS(1)
 
+/**
+ * @brief Used for debugging to print ip addresses.
+ *
+ * @param addr sockaddr for socket using ipv4 or ipv6.
+ * @return char*
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.h:37]: The output should have some description, not only the type.

File: subsys/net/lib/lwm2m/lwm2m_engine.h [hunk 2]
```diff
@@ -30,63 +30,286 @@
...
+ */
 char *lwm2m_sprint_ip_addr(const struct sockaddr *addr);
+
+/**
+ * @brief Converts the token to a printable format.
+ *
+ * @param[in] token Token to be printed.
+ * @param tkl Lenghts of token.
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.h:45]: Please be consistent, either document whether the parameter is input or output for all parameters, or none.

File: subsys/net/lib/lwm2m/lwm2m_engine.h [hunk 3]
```diff
@@ -30,63 +30,286 @@
...
-
-int lwm2m_get_path_reference_ptr(struct lwm2m_engine_obj *obj, struct lwm2m_obj_path *path,
-				 void **ref);
-
 int lwm2m_perform_composite_observation_op(struct lwm2m_message *msg, uint8_t *token,
 					   uint8_t token_length, sys_slist_t *lwm2m_path_list);
 
+/**
+ * @brief Checks if context has access to the object specified by @p path
+ * Returns false if the bootstrap flag of @p client_ctx is not set.
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.h:88]: Shouln't this be documented as a part of `@return `?

File: subsys/net/lib/lwm2m/lwm2m_engine.h [hunk 4]
```diff
@@ -30,63 +30,286 @@
...
+ *
+ * @param[in] obj_inst_id Object instance id of the firmware object instance.
+ * @param[in] result (RESULT_DEFAULT through RESULT_UNSUP_PROTO)
+ */
 void lwm2m_firmware_set_update_result_inst(uint16_t obj_inst_id, uint8_t result);
+
+/**
+ * @brief Equivalent to lwm2m_firmware_set_update_state_inst(0, state).
+ *
+ * @param[in] state
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.h:195]: Description is missing.

File: subsys/net/lib/lwm2m/lwm2m_engine.h [hunk 5]
```diff
@@ -30,63 +30,286 @@
...
+ */
 void lwm2m_firmware_set_update_state(uint8_t state);
+
+/**
+ * @brief Equivalent to lwm2m_firmware_set_update_result_inst(0, result).
+ *
+ * @param[in] result
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.h:202]: Description is missing.

File: subsys/net/lib/lwm2m/lwm2m_engine.h [hunk 6]
```diff
@@ -30,63 +30,286 @@
...
+ * (i.e Gets the value of resource 5/ @p obj_inst_id /3)
+ * @param[in] obj_inst_id
+ * @return (STATE_IDLE through STATE_UPDATING)
+ */
 uint8_t lwm2m_firmware_get_update_state_inst(uint16_t obj_inst_id);
+
+/**
+ * @brief Equivalent to lwm2m_firmware_get_update_state_inst(0).
+ *
+ * @return uint8_t
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.h:218]: Description is missing.

File: subsys/net/lib/lwm2m/lwm2m_engine.h [hunk 7]
```diff
@@ -30,63 +30,286 @@
...
+ *
+ * @param[in] result
+ */
 void lwm2m_firmware_set_update_result(uint8_t result);
+
+/**
+ * @brief Returns the update state (as specified in  LWM2M SPEC E.6 regarding the firmware update)
+ * of the Firmware object instance given by @p obj_inst_id.
+ * (i.e Gets the value of resource 5/ @p obj_inst_id /3)
+ * @param[in] obj_inst_id
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.h:210]: Description is missing.

File: subsys/net/lib/lwm2m/lwm2m_engine.h [hunk 8]
```diff
@@ -30,63 +30,286 @@
...
+ *
+ * @return uint8_t
+ */
 uint8_t lwm2m_firmware_get_update_state(void);
+
+/**
+ * @brief Returns the result state (as specified in  LWM2M SPEC E.6 regarding the firmware update)
+ * of the Firmware object instance given by @p obj_inst_id.
+ * (i.e Gets the value of resource 5/ @p obj_inst_id /5)
+ * @param[in] obj_inst_id
```
Comment (rlubos) [inline subsys/net/lib/lwm2m/lwm2m_engine.h:226]: Description is missing. There are a few more places like this, I hope you see what I mean, I don't want to spam too much.

File: subsys/net/lib/lwm2m/lwm2m_registry.h [hunk 9]
```diff
@@ -6,45 +6,191 @@
...
+ */
 bool lwm2m_engine_shall_report_obj_version(const struct lwm2m_engine_obj *obj);
 
-int lwm2m_engine_get_resource(const char *pathstr, struct lwm2m_engine_res **res);
-
+/**
+ * @brief Returns the binding mode in use. Defaults to UDP (U).
+ *
+ * @param[out] binding Data buffer.
+ */
```
Comment (SeppoTakalo) [inline subsys/net/lib/lwm2m/lwm2m_registry.h:171]:
```suggestion
/**
 * @brief Returns the binding mode in use.
 *
 * Defaults to UDP (U). In LwM2M 1.0 when queue mode is enabled, return "UQ".
 *
 * @param[out] binding Data buffer.
 */
```

**Review by SeppoTakalo — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**